collector/collector_test.go

```diff
@@ -0,0 +1,344 @@
+
+		col := NewCollector(log.NewJSONLogger(os.Stdout), &Config{})
+		col.openDatabase = func(_ string) (*sql.DB, error) { return db, nil }
+
+		// No metrics should be collected; An empty buffer should match the output
+		// of the scraped metrics
+		err := testutil.CollectAndCompare(col, &bytes.Buffer{})
+		require.NoError(t, err)
+
+		// Checks that db is closed and all queries are called
+		require.NoError(t, mock.ExpectationsWereMet())
+	})
+
+	t.Run("Database connection fails", func(t *testing.T) {
+		col := NewCollector(log.NewJSONLogger(os.Stdout), &Config{})
+
+		openErr := errors.New("failed to open database")
+		col.openDatabase = func(_ string) (*sql.DB, error) { return nil, openErr }
+
+		// No metrics should be scraped if the database fails to open
+		err := testutil.CollectAndCompare(col, &bytes.Buffer{})
+		require.NoError(t, err)
+	})
+}
+
+func TestCollector_collectStorageMetrics(t *testing.T) {
+	t.Run("Row error", func(t *testing.T) {
+		db, mock, err := sqlmock.New(sqlmock.QueryMatcherOption(sqlmock.QueryMatcherEqual))
+		require.NoError(t, err)
+
+		rowErr := errors.New("failed for inexplicable reasons")
+
+		mock.ExpectQuery(storageMetricQuery).
+			WillReturnRows(
+				sqlmock.NewRows([]string{"STORAGE_BYTES", "STAGE_BYTES", "FAILSAFE_BYTES"}).AddRow(
+					sql.NullString{}, sql.NullString{}, sql.NullString{},
+				).RowError(0, rowErr),
+			).
+			RowsWillBeClosed()
+
+		col := NewCollector(log.NewJSONLogger(os.Stdout), &Config{})
+		col.openDatabase = func(_ string) (*sql.DB, error) { return db, nil }
+
+		metricChan := make(chan prometheus.Metric, 3)
+		collectionDoneChan := make(chan struct{})
+
+		go func() {
+			err := col.collectStorageMetrics(db, metricChan)
+			require.ErrorContains(t, err, "failed to fetch row:")
+			require.ErrorIs(t, err, rowErr)
+			close(collectionDoneChan)
+		}()
+
+		select {
+		case <-collectionDoneChan: // OK
+		case <-time.After(time.Second):
+			require.Fail(t, "Timed out waiting for collectStorageMetrics to complete")
+		}
+	})
+
+	t.Run("No rows returned", func(t *testing.T) {
+		db, mock, err := sqlmock.New(sqlmock.QueryMatcherOption(sqlmock.QueryMatcherEqual))
+		require.NoError(t, err)
+
+		mock.ExpectQuery(storageMetricQuery).
+			WillReturnRows(
+				sqlmock.NewRows([]string{"STORAGE_BYTES", "STAGE_BYTES", "FAILSAFE_BYTES"}),
+			).
+			RowsWillBeClosed()
+
+		col := NewCollector(log.NewJSONLogger(os.Stdout), &Config{})
+		col.openDatabase = func(_ string) (*sql.DB, error) { return db, nil }
+
+		metricChan := make(chan prometheus.Metric, 3)
+		collectionDoneChan := make(chan struct{})
+
+		go func() {
+			err := col.collectStorageMetrics(db, metricChan)
+			require.Equal(t, "expected a single row to be returned, but none was found", err.Error())
+			close(collectionDoneChan)
+		}()
+
+		select {
+		case <-collectionDoneChan: // OK
+		case <-time.After(time.Second):
+			require.Fail(t, "Time out waiting for collectStorageMetrics to complete")
+		}
```

Comment:
Could this be replaced with a [require.Eventually](https://pkg.go.dev/github.com/stretchr/testify/require?utm_source=godoc#Eventually) where the conditional is something like 

```go
func () bool {
  <-collectionDoneChan
  return true
}
```

Reply:
I played around with this a bit, I don't think it's the right option here. I think `require.Eventually` is more relevant when the function we pass in is non-blocking (the function is meant to be a polling function); Otherwise it spins up a goroutine each tick interval, and having to specify a tick interval at all seems weird in this instance. 

This is something that's common enough that you'd think there'd be a specialty function for this, but that doesn't seem to be the case?